### PR TITLE
Logic overhaul

### LINKED ISSUE #31: Support sibling modules with import stack tracing
## Is your feature request related to a problem? Please describe.

This feature request is a follow on to #27. [The initial implementation](https://github.com/IBM/import-tracker/pull/30) is a best-effort and doesn't handle the case where a downstream import comes in via a sibling module (e.g. `submod2` and `submod3` in the [sample `inter_mod_deps` library](https://github.com/IBM/import-tracker/tree/main/test/sample_libs/inter_mod_deps).

The request here is to patch the issues discussed in the WIP solution: https://github.com/IBM/import-tracker/commit/3ed538bc122d0540fa92202e2ff75d8f6d91dcaa

## Describe the solution you'd like

All stack lines involved in attributing a given import to a given module should be captured!

## Additional context

An incomplete solution has been started on a branch: https://github.com/IBM/import-tracker/tree/save/StackTrackingDeferredSiblings

### PR DESCRIPTION
**Pull request by gabe-l-hart**
## Description

This is a major change! It fundamentally rewrites the core logic for tracking imports and rearranges the arguments to the import tracking functionality. The high-level gist is:

* Rather than capturing imports during the processing of an `import_module`, the imports are computed after importing the target module by recursively inspecting the bytecode for all modules stemming from the target.
* The tracking no longer needs to launch `subprocesses` to perform recursion because it does not rely on the diff in `sys.modules`
* It's _way_ faster!

## But why?

Ok, the old way was working pretty well, so why refactor it all? The obvious answer is `speed`, but the less obvious answer is actually the correct one: the old implementation was not answering the right question. The old implementation answered the question

```
What modules are brought into sys.modules between starting the import of <target> and concluding the import of <target>?
```

Instead, what we _really_ want to know is:

```
If we stripped away all code not required for <target>, what modules would we need to have installed for the import of <target> to work?
```

The difference here comes down to whether you count siblings of nested dependencies. This is much easier to describe with an example:

```
deep_siblings/
├── __init__.py
├── blocks
│   ├── __init__.py
│   ├── bar_type
│   │   ├── __init__.py
│   │   └── bar.py # imports alog
│   └── foo_type
│       ├── __init__.py
│       └── foo.py # imports yaml
└── workflows
    ├── __init__.py
    └── foo_type
        ├── __init__.py
        └── foo.py # imports ..blocks.foo_type.foo
```

In this example, under the old implementation, `workflows.foo_type.foo` would depend on both `alog` and `yaml` because the `..blocks` portion of the import requires that _all_ of the dependencies of `blocks` be brought into `sys.modules`. This, however, voids the value proposition of finding separable import sets. Under the new implementation, `workflows.foo_type.foo` only depends on `yaml` because it imports `blocks.foo_type.foo` from the deepest point where the only requirement is `yaml`.

## Testing

This PR also refactors all of the tests since it's now more practical to test the import logic by calling `track_module` directly rather than simulating a call to `main`. With the exception of a few tests that have been removed due to their irrelevance with the removal of `side_effects_modules` (that went away!), all tests have been maintained and updated. There are several where the results change, but in each case the change is actually an improvement on a known corner-case, with one specific exception: `LazyModule` (more on that below).

## What breaks in the API?

* The `side_effects_modules` argument is gone. This was a hack to work around the fact that there were some modules that, when trapped by a `DeferredModule` would cause the overall import to fail. With the refactor, this is unnecessary as the import proceeds exactly as normal with no interferance.
* The output with `track_import_stacks` is different. It no longer attempts to look like stack traces, but it is actually more useful. Now, instead of a partially-useful stack trace, it's a list of lists where each entry is stack of module imports that causes the given dependency allocation. This fixes #31!
* By default, `import_module` stops looking for imports at the boundary of the target module's parent library. This means that if a third party module transitively imports another third party module, it won't be allocated to the target _unless_ `full_depth=True` is given
* `LazyModule` is gone! This tool was a bit of a hack anyway and is no longer necessary

### CODE REVIEW

File: test/test_import_tracker.py [hunk 1]
```diff
@@ -73,22 +69,330 @@ def test_track_module_recursive():
     }
 
 
-def test_track_module_with_log_level():
```
Comment (gabe-l-hart) [inline test/test_import_tracker.py:76]: `log_level` is no longer configured in a subprocess, so that argument is gone

File: test/test_import_tracker.py [hunk 2]
```diff
@@ -73,22 +69,330 @@ def test_track_module_recursive():
...
-        recursive=True,
         submodules=["sample_lib.submod1"],
     )
     assert sample_lib_mapping == {
         "sample_lib": sorted(["conditional_deps", "alog", "yaml"]),
         "sample_lib.submod1": ["conditional_deps"],
     }
+
+
+def test_sibling_import():
```
Comment (gabe-l-hart) [inline test/test_import_tracker.py:84]: These tests moved over from `test_main`

File: test/test_main.py [hunk 3]
```diff
@@ -140,306 +138,3 @@ def test_lib_with_lazy_imports(capsys):
     assert captured.out
     parsed_out = json.loads(captured.out)
     assert "lazy_import_errors" in parsed_out
-
-
-def test_lib_with_side_effect_imports():
```
Comment (gabe-l-hart) [inline test/test_main.py:145]: These tests all moved over to `test_import_tracker`

File: import_tracker/setup_tools.py [hunk 4]
```diff
@@ -66,7 +66,7 @@ def parse_requirements(
     # Get the set of required modules for each of the listed extras modules
     library_import_mapping = track_module(
         library_name,
-        recursive=True,
```
[the remark this answers is not in the document]
Comment (DoraAgali) [reply import_tracker/setup_tools.py:69]: hmm, I think yes

[the remark this answers is not in the document]
Comment (gabe-l-hart) [reply import_tracker/setup_tools.py:69]: After some local testing, I decided to flip the default to `full_depth=True` for parse_requirements, but still allow it to be set directly

**Comment by gabe-l-hart** [comment]
@DoraAgali @gkumbhat The last open question with this is whether to fully remove `LazyModule`. I'm inclined to remove it since it was only added to support "hiding" imports in cross-cutting modules and is not used elsewhere. With this refactor, that hiding should not be necessary anymore as long as imports in those cross-cutting modules are managed correctly.

File: import_tracker/__main__.py [hunk 5]
```diff
@@ -568,39 +50,12 @@ def main():
...
-    )
-    parser.add_argument(
-        "--num_jobs",
-        "-j",
-        type=int,
-        help="Number of workers to spawn when recursing",
-        default=0,
-    )
-    parser.add_argument(
-        "--side_effect_modules",
```
Comment (gkumbhat) [inline import_tracker/__main__.py:599]: my nit would be to raise error if someone passes these, to make sure existing users address things with new update
